src/audio/mfcc/mfcc_hifi4.c

```diff
@@ -0,0 +1,289 @@
+	int i;
+
+	/* Set buf as circular buffer 0 */
+	set_circular_buf0(buf->addr, buf->end_addr);
+
+	/* very strange this align load is unexpected
+	 * so use load a 32bit to replace 16x4 align load.
+	 */
+	for (i = 0; i < n; i++) {
+		AE_L32_XC(in_sample, in, inc);
+		/* sizeof(ae_int32) = 4 */
+		AE_S32_L_IP(in_sample, out, 4);
+	}
+	if (prev_data_length & 0x01) {
+		AE_L16_XC(sample, (ae_int16 *)in, sizeof(ae_int16));
+		AE_S16_0_IP(sample, (ae_int16 *)out, sizeof(ae_int16));
+	}
+
+	buf->s_avail -= prev_data_length;
+	buf->s_free += prev_data_length;
+	buf->r_ptr = in;
+}
+
+void mfcc_fill_fft_buffer(struct mfcc_state *state)
+{
+	struct mfcc_buffer *buf = &state->buf;
+	struct mfcc_fft *fft = &state->fft;
+	int idx = fft->fft_fill_start_idx;
+	ae_int16 *out = (ae_int16 *)&fft->fft_buf[idx].real;
+	ae_int16 *in = (ae_int16 *)state->prev_data;
+	ae_int16x4 sample;
+	const int buf_inc = sizeof(ae_int16);
+	const int fft_inc = sizeof(fft->fft_buf[0]);
+	int j;
+
+	/* Copy overlapped samples from state buffer. Imaginary part of input
+	 * remains zero.
+	 */
+	for (j = 0; j < state->prev_data_size; j++) {
+		AE_L16_XP(sample, in, buf_inc);
+		AE_S16_0_XP(sample, out, fft_inc);
+	}
+
+	/* Copy hop size of new data from circular buffer */
+	idx += state->prev_data_size;
+	in = (ae_int16 *)buf->r_ptr;
+	out = (ae_int16 *)&fft->fft_buf[idx].real;
+	set_circular_buf1(buf->addr, buf->end_addr);
+	for (j = 0; j < fft->fft_hop_size; j++) {
+		AE_L16_XC(sample, in, buf_inc);
+		AE_S16_0_XP(sample, out, fft_inc);
+	}
+
+	buf->s_avail -= fft->fft_hop_size;
+	buf->s_free += fft->fft_hop_size;
+	buf->r_ptr = (int16_t *)in;
+
+	/* Copy for next time data back to overlap buffer */
+	idx = fft->fft_fill_start_idx + fft->fft_hop_size;
+	in = (ae_int16 *)&fft->fft_buf[idx].real;
+	out = (ae_int16 *)state->prev_data;
+	for (j = 0; j < state->prev_data_size; j++) {
+		AE_L16_XP(sample, in, fft_inc);
+		AE_S16_0_XP(sample, out, buf_inc);
+	}
+}
+
+#ifdef MFCC_NORMALIZE_FFT
+int mfcc_normalize_fft_buffer(struct mfcc_state *state)
+{
+	struct mfcc_fft *fft = &state->fft;
+	ae_p16s *in = (ae_p16s *)&fft->fft_buf[fft->fft_fill_start_idx].real;
+	ae_int32x2 sample;
+	ae_int32x2 max = AE_ZERO32();
+	const int fft_inc = sizeof(fft->fft_buf[0]);
+	int shift;
+	int j;
+
+	for (j = 0; j < fft->fft_size; j++) {
+		/* load 16-bit data to middle of 32-bit container*/
+		AE_L16M_XU(sample, in, fft_inc);
+		max = AE_MAXABS32S(max, sample);
+	}
+	shift = AE_NSAZ32_L(max) - 8;/* 16 bit data */
+	shift = MAX(shift, 0);
+	shift = MIN(shift, MFCC_NORMALIZE_MAX_SHIFT);
+	return shift;
+}
+#endif
+
+void mfcc_apply_window(struct mfcc_state *state, int input_shift)
+{
+	struct mfcc_fft *fft = &state->fft;
+	const int fft_inc = sizeof(fft->fft_buf[0]);
+	ae_int16 *win_in = (ae_int16 *)state->window;
+	const int win_inc = sizeof(ae_int16);
+	ae_int32x2 temp;
+	ae_int16x4 win;
+	int j;
+
+#if MFCC_FFT_BITS == 16
+	/* TODO: Use proper multiply and saturate function to make sure no overflows */
```

Comment:
You can drop these comments about saturate, it's now handled with AE multiply instructions.